In [76]:
import pandas as pd
import numpy as np

In [77]:
Data = pd.read_csv('House_Price_Regression.csv')

np.random.seed(123)

Data['type'] = np.random.uniform(low=0, high=3, size=len(Data)).round()

Data['rating'] = np.random.uniform(low=0, high=5, size=len(Data)).round()

Data.head()

neighborhood_recode   latitude  longitude    price  no_of_bedrooms  \
0                 46.0  25.113208  55.138932  2700000               1   
1                 46.0  25.106809  55.151201  2850000               2   
2                 36.0  25.063302  55.137728  1150000               3   
3                 11.0  25.227295  55.341761  2850000               2   
4                 46.0  25.114275  55.139764  1729200               0   

   no_of_bathrooms  quality_recode  maid_room_recode  unfurnished_recode  \
0                2             2.0               0.0                 0.0   
1                2             2.0               0.0                 0.0   
2                5             2.0               1.0                 1.0   
3                3             1.0               0.0                 1.0   
4                1             2.0               0.0                 0.0   

   balcony_recode  ...  private_jacuzzi_recode  private_pool_recode  \
0             1.0  ...                     0.0                  0.0   
1             1.0  ...                     0.0                  0.0   
2             1.0  ...                     1.0                  0.0   
3             1.0  ...                     0.0                  0.0   
4             0.0  ...                     0.0                  0.0   

   security_recode  shared_gym_recode  shared_pool_recode  shared_spa_recode  \
0              0.0                1.0                 0.0                0.0   
1              0.0                1.0                 1.0                0.0   
2              1.0                1.0                 1.0                0.0   
3              0.0                0.0                 0.0                0.0   
4              1.0                1.0                 1.0                1.0   

   view_of_water_recode  size_in_m_2  type  rating  
0                   1.0   100.242337   2.0     3.0  
1                   1.0   146.972546   1.0     2.0  
2                   1.0   181.253753   1.0     5.0  
3                   0.0   187.664060   2.0     5.0  
4                   1.0    47.101821   2.0     1.0  

[5 rows x 31 columns]

In [78]:
Data_quant = Data.loc[: , ['latitude', 'longitude', 'price', 'size_in_m_2', 'no_of_bedrooms', 'no_of_bathrooms']]

In [79]:
Data_quant.head()

latitude  longitude    price  size_in_m_2  no_of_bedrooms  no_of_bathrooms
0  25.113208  55.138932  2700000   100.242337               1                2
1  25.106809  55.151201  2850000   146.972546               2                2
2  25.063302  55.137728  1150000   181.253753               3                5
3  25.227295  55.341761  2850000   187.664060               2                3
4  25.114275  55.139764  1729200    47.101821               0                1

In [80]:
Data_binary = Data.loc[: , ['private_garden_recode', 'private_gym_recode', 'private_jacuzzi_recode', 'balcony_recode']]

Data_binary.head()

private_garden_recode  private_gym_recode  private_jacuzzi_recode  \
0                    0.0                 0.0                     0.0   
1                    0.0                 0.0                     0.0   
2                    0.0                 0.0                     1.0   
3                    0.0                 0.0                     0.0   
4                    0.0                 0.0                     0.0   

   balcony_recode  
0             1.0  
1             1.0  
2             1.0  
3             1.0  
4             0.0

In [81]:
Data_multiclass = Data.loc[: , ['quality_recode', 'rating', 'type']]

Data_multiclass.head()

quality_recode  rating  type
0             2.0     3.0   2.0
1             2.0     2.0   1.0
2             2.0     5.0   1.0
3             1.0     5.0   2.0
4             2.0     1.0   2.0

In [82]:
Data_mixed = Data.loc[: , ['latitude', 'longitude', 'price', 'size_in_m_2', 'no_of_bedrooms', 'no_of_bathrooms','private_garden_recode', 'private_gym_recode', 'private_jacuzzi_recode', 'balcony_recode','quality_recode', 'rating', 'type']]

Data_mixed.head()

latitude  longitude    price  size_in_m_2  no_of_bedrooms  \
0  25.113208  55.138932  2700000   100.242337               1   
1  25.106809  55.151201  2850000   146.972546               2   
2  25.063302  55.137728  1150000   181.253753               3   
3  25.227295  55.341761  2850000   187.664060               2   
4  25.114275  55.139764  1729200    47.101821               0   

   no_of_bathrooms  private_garden_recode  private_gym_recode  \
0                2                    0.0                 0.0   
1                2                    0.0                 0.0   
2                5                    0.0                 0.0   
3                3                    0.0                 0.0   
4                1                    0.0                 0.0   

   private_jacuzzi_recode  balcony_recode  quality_recode  rating  type  
0                     0.0             1.0             2.0     3.0   2.0  
1                     0.0             1.0             2.0     2.0   1.0  
2                     1.0             1.0             2.0     5.0   1.0  
3                     0.0             1.0             1.0     5.0   2.0  
4                     0.0             0.0             2.0     1.0   2.0

## Distancia Euclidea

In [83]:
def Dist_Euclidea(x_i, x_r):

        Dist_Euclidea = ( ( x_i - x_r )**2 ).sum()

        Dist_Euclidea = np.sqrt(Dist_Euclidea)

        return Dist_Euclidea

In [84]:
Dist_Euclidea(Data_quant.iloc[2,:] , Data_quant.iloc[5,:])

1969900.0019225744

In [85]:
def Matrix_Dist_Euclidea(Data):

    Data = Data.to_numpy()

    n = len(Data)

    M =  np.empty((n , n))

    
    for i in range(0, n):

         for r in range(0, n):

             if i >= r :
               
                 M[i,r] = 0

             else :

                 M[i,r] = Dist_Euclidea(Data[i,:] , Data[r,:])   

                      
    return M 

In [86]:
M_Euclidean = Matrix_Dist_Euclidea(Data=Data_quant)

In [87]:
M_Euclidean

array([[      0.        ,  150000.00728238, 1550000.00212124, ...,
         200000.01565902, 2025000.00024491, 1939113.00017324],
       [      0.        ,       0.        , 1700000.00034859, ...,
          50000.01055292, 2175000.00140691, 2089113.00126347],
       [      0.        ,       0.        ,       0.        , ...,
        1750000.00000111,  475000.01333737,  389113.0147095 ],
       ...,
       [      0.        ,       0.        ,       0.        , ...,
              0.        , 2225000.00274952, 2139113.00257909],
       [      0.        ,       0.        ,       0.        , ...,
              0.        ,       0.        ,   85887.00018092],
       [      0.        ,       0.        ,       0.        , ...,
              0.        ,       0.        ,       0.        ]])

In [88]:
M_Euclidean = M_Euclidean + M_Euclidean.T

M_Euclidean

array([[      0.        ,  150000.00728238, 1550000.00212124, ...,
         200000.01565902, 2025000.00024491, 1939113.00017324],
       [ 150000.00728238,       0.        , 1700000.00034859, ...,
          50000.01055292, 2175000.00140691, 2089113.00126347],
       [1550000.00212124, 1700000.00034859,       0.        , ...,
        1750000.00000111,  475000.01333737,  389113.0147095 ],
       ...,
       [ 200000.01565902,   50000.01055292, 1750000.00000111, ...,
              0.        , 2225000.00274952, 2139113.00257909],
       [2025000.00024491, 2175000.00140691,  475000.01333737, ...,
        2225000.00274952,       0.        ,   85887.00018092],
       [1939113.00017324, 2089113.00126347,  389113.0147095 , ...,
        2139113.00257909,   85887.00018092,       0.        ]])

## Distancia de Minkowski

In [89]:
def Dist_Minkowski(x_i, x_r, q):

    Dist_Minkowski = ( ( ( abs( x_i - x_r) )**q ).sum() )**(1/q)

    return Dist_Minkowski

In [90]:
Dist_Minkowski(Data_quant.iloc[2,:], Data_quant.iloc[5,:], q=1)

1969992.0102169998

In [91]:
Dist_Minkowski(Data_quant.iloc[2,:], Data_quant.iloc[5,:], q=2)

1969900.0019225744

In [92]:
Dist_Minkowski(Data_quant.iloc[2,:], Data_quant.iloc[5,:], q=3)

1969900.000000055

In [93]:
Dist_Minkowski(Data_quant.iloc[2,:], Data_quant.iloc[5,:], q=10)

1969900.0000000016

In [94]:
def Matrix_Dist_Minkowski(Data, q):

    Data = Data.to_numpy()

    n = len(Data)

    M =  np.empty((n , n))

    
    for i in range(0, n):

         for r in range(0, n):

             if i >= r :
               
                 M[i,r] = 0

             else :

                 M[i,r] = Dist_Minkowski(Data[i,:] , Data[r,:], q)   
                      
    return M 

In [95]:
M_Minkowski = Matrix_Dist_Minkowski(Data=Data_quant, q=1)

M_Minkowski

array([[      0.      ,  150047.748877, 1550086.062526, ...,
         200084.359869, 2025031.624379, 1939138.969796],
       [      0.      ,       0.      , 1700038.338187, ...,
          50036.62379 , 2175079.33592 , 2089186.681337],
       [      0.      ,       0.      ,       0.      , ...,
        1750002.301489,  475117.608205,  389224.964166],
       ...,
       [      0.      ,       0.      ,       0.      , ...,
              0.      , 2225115.802424, 2139223.298103],
       [      0.      ,       0.      ,       0.      , ...,
              0.      ,       0.      ,   85892.654583],
       [      0.      ,       0.      ,       0.      , ...,
              0.      ,       0.      ,       0.      ]])

In [96]:
M_Minkowski = M_Minkowski + M_Minkowski.T

M_Minkowski

array([[      0.      ,  150047.748877, 1550086.062526, ...,
         200084.359869, 2025031.624379, 1939138.969796],
       [ 150047.748877,       0.      , 1700038.338187, ...,
          50036.62379 , 2175079.33592 , 2089186.681337],
       [1550086.062526, 1700038.338187,       0.      , ...,
        1750002.301489,  475117.608205,  389224.964166],
       ...,
       [ 200084.359869,   50036.62379 , 1750002.301489, ...,
              0.      , 2225115.802424, 2139223.298103],
       [2025031.624379, 2175079.33592 ,  475117.608205, ...,
        2225115.802424,       0.      ,   85892.654583],
       [1939138.969796, 2089186.681337,  389224.964166, ...,
        2139223.298103,   85892.654583,       0.      ]])

## Distancia Canberra

In [97]:
def Dist_Canberra(x_i, x_r):

    numerator =  abs( x_i - x_r )

    denominator =  ( abs(x_i) + abs(x_r) )
       
    numerator=np.array([numerator], dtype=float)

    denominator=np.array([denominator], dtype=float)

    # The following code is to eliminate zero division problems

    Dist_Canberra = ( np.divide( numerator , denominator , out=np.zeros_like(numerator), where=denominator!=0) ).sum() 

    return Dist_Canberra

In [98]:
Dist_Canberra(Data.iloc[2,:] , Data.iloc[5,:])

8.8284791957755

In [99]:
def Matrix_Dist_Canberra(Data):

    Data = Data.to_numpy()

    n = len(Data)

    M =  np.empty((n , n))

    
    for i in range(0, n):

         for r in range(0, n):

             if i >= r :
               
                 M[i,r] = 0

             else :

                 M[i,r] = Dist_Canberra(Data[i,:] , Data[r,:])   
                     
    return M 

In [100]:
M_Canberra = Matrix_Dist_Canberra(Data=Data_quant)

M_Canberra

array([[0.        , 0.5496257 , 1.61996314, ..., 1.25082356, 0.78797391,
        0.70959891],
       [0.        , 0.        , 1.15900459, ..., 0.74009173, 1.31434216,
        1.24077176],
       [0.        , 0.        , 0.        , ..., 0.44223489, 1.63990915,
        1.55106392],
       ...,
       [0.        , 0.        , 0.        , ..., 0.        , 2.00032192,
        1.93106183],
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        0.09956138],
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        0.        ]])

In [101]:
M_Canberra = M_Canberra + M_Canberra.T

M_Canberra

array([[0.        , 0.5496257 , 1.61996314, ..., 1.25082356, 0.78797391,
        0.70959891],
       [0.5496257 , 0.        , 1.15900459, ..., 0.74009173, 1.31434216,
        1.24077176],
       [1.61996314, 1.15900459, 0.        , ..., 0.44223489, 1.63990915,
        1.55106392],
       ...,
       [1.25082356, 0.74009173, 0.44223489, ..., 0.        , 2.00032192,
        1.93106183],
       [0.78797391, 1.31434216, 1.63990915, ..., 2.00032192, 0.        ,
        0.09956138],
       [0.70959891, 1.24077176, 1.55106392, ..., 1.93106183, 0.09956138,
        0.        ]])

## Distancia Pearson

In [102]:
def Dist_Pearson(x_i, x_r, variance) :

    Dist_Pearson = ( ( x_i - x_r )**2 / variance ).sum()

    Dist_Pearson = np.sqrt(Dist_Pearson)

    return Dist_Pearson

In [103]:
Dist_Pearson(x_i=Data_quant.iloc[2,:], x_r=Data_quant.iloc[5,:], variance = np.var(Data_quant , axis=0, ddof=1))

3.8239347754243425

In [104]:
def Matrix_Dist_Pearson(Data):

    Data = Data.to_numpy()

    n = len(Data)

    M =  np.empty((n , n))

    
    for i in range(0, n):

         for r in range(0, n):

             if i >= r :
               
                 M[i,r] = 0

             else :

                 M[i,r] = Dist_Pearson(Data[i,:] , Data[r,:], variance=np.var(Data, axis=0, ddof=1))   
                     
    return M 

In [105]:
M_Pearson = Matrix_Dist_Pearson(Data=Data_quant)

M_Pearson

array([[0.        , 1.21345279, 3.77819153, ..., 4.95085256, 1.66393723,
        0.94314867],
       [0.        , 0.        , 3.17880538, ..., 4.43821303, 2.03523139,
        1.60958264],
       [0.        , 0.        , 0.        , ..., 3.82999056, 4.01162596,
        3.76955491],
       ...,
       [0.        , 0.        , 0.        , ..., 0.        , 4.69603617,
        5.13174853],
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        1.09780835],
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        0.        ]])

In [106]:
M_Pearson = M_Pearson + M_Pearson.T

M_Pearson

array([[0.        , 1.21345279, 3.77819153, ..., 4.95085256, 1.66393723,
        0.94314867],
       [1.21345279, 0.        , 3.17880538, ..., 4.43821303, 2.03523139,
        1.60958264],
       [3.77819153, 3.17880538, 0.        , ..., 3.82999056, 4.01162596,
        3.76955491],
       ...,
       [4.95085256, 4.43821303, 3.82999056, ..., 0.        , 4.69603617,
        5.13174853],
       [1.66393723, 2.03523139, 4.01162596, ..., 4.69603617, 0.        ,
        1.09780835],
       [0.94314867, 1.60958264, 3.76955491, ..., 5.13174853, 1.09780835,
        0.        ]])

## Distancia de Mahalanobis

In [107]:
def Dist_Mahalanobis_1(x_i, x_r, Data):

    # All the columns of Data must be type = 'float' or 'int' (specially not 'object'), in other case we will find 
    # dimensional problems when Python compute   x @ S_inv @ x.T

    x = (x_i - x_r)

    x = np.array([x]) # necessary step to transpose a 1D array

    S_inv = np.linalg.inv( np.cov(Data , rowvar=False) ) # inverse of covariance matrix

    Dist_Maha = np.sqrt( x @ S_inv @ x.T )  # x @ S_inv @ x.T = np.matmul( np.matmul(x , S_inv) , x.T )

    Dist_Maha = float(Dist_Maha)

    return Dist_Maha

In [108]:
Dist_Mahalanobis_1(Data_quant.iloc[2,:] , Data_quant.iloc[5,:] , Data=Data_quant)

3.931396144771864

In [109]:
def Dist_Mahalanobis_2(x_i, x_r, S_inv):  # Más eficiente que la anterior

    # All the columns of Data must be type = 'float' or 'int' (specially not 'object'), in other case we will find 
    # dimensional problems when Python compute   x @ S_inv @ x.T

    x = x_i - x_r

    x = np.array([x]) # necessary step to transpose a 1D array

    Dist_Maha = np.sqrt( x @ S_inv @ x.T )  # x @ S_inv @ x.T = np.matmul( np.matmul(x , S_inv) , x.T )

    Dist_Maha = float(Dist_Maha)

    return Dist_Maha

In [110]:
Dist_Mahalanobis_2(x_i = Data_quant.iloc[2,:] , x_r = Data_quant.iloc[5,:] , S_inv = np.linalg.inv( np.cov(Data_quant, rowvar=False) ) )

3.931396144771864

In [111]:
def Dist_Mahalanobis_3(x, S_inv):  # Más eficiente que la anterior

    # All the columns of Data must be type = 'float' or 'int' (specially not 'object'), in other case we will find 
    # dimensional problems when Python compute   x @ S_inv @ x.T

    # x = (x_i - x_r)

    # x = np.array([x]) # necessary step to transpose a 1D array

    Dist_Maha = np.sqrt( x @ S_inv @ x.T )  # x @ S_inv @ x.T = np.matmul( np.matmul(x , S_inv) , x.T )

    Dist_Maha = float(Dist_Maha)

    return Dist_Maha

In [112]:
Dist_Mahalanobis_3(x = Data_quant.iloc[2,:] - Data_quant.iloc[5,:] , S_inv=np.linalg.inv( np.cov(Data_quant, rowvar=False) ))

3.931396144771864

In [113]:
def Matrix_Dist_Mahalanobis_1(Data):

    Data = Data.to_numpy()

    n = len(Data)

    M =  np.empty((n , n))

   
    for i in range(0, n):

         for r in range(0, n):

             if i >= r :
               
                 M[i,r] = 0

             else :

                 M[i,r] = Dist_Mahalanobis_1(x_i = Data[i,:] , x_r = Data[r,:] , Data=Data )

                      
    return M 

In [114]:
Matrix_Dist_Mahalanobis_1(Data=Data_quant.iloc[0:1000, :]) # 2.25 mins

KeyboardInterrupt: 

In [ ]:
Matrix_Dist_Mahalanobis_1(Data=Data_quant) # 8.52 mins

array([[0.        , 2.11289055, 3.7975463 , ..., 4.51559865, 2.31688444,
        1.10588047],
       [0.        , 0.        , 4.35615967, ..., 4.93340427, 2.74011739,
        2.12938584],
       [0.        , 0.        , 0.        , ..., 3.17779509, 3.49432487,
        3.23723317],
       ...,
       [0.        , 0.        , 0.        , ..., 0.        , 3.58695453,
        4.11275247],
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        1.46894947],
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        0.        ]])

In [115]:
def Matrix_Dist_Mahalanobis_2(Data):

    Data = Data.to_numpy()

    n = len(Data)

    M =  np.empty((n , n))

    S_inv = np.linalg.inv( np.cov(Data , rowvar=False) )

   
    for i in range(0, n):

         for r in range(0, n):

             if i >= r :
               
                 M[i,r] = 0

             else :

                 M[i,r] = Dist_Mahalanobis_2(x_i = Data[i,:] , x_r = Data[r,:] , S_inv = S_inv  )

                      
    return M 

In [ ]:
Matrix_Dist_Mahalanobis_2(Data=Data_quant.iloc[0:1000, :]) # 6.8 seg

array([[0.        , 2.16321817, 3.81131086, ..., 2.74083537, 3.84799958,
        1.06041471],
       [0.        , 0.        , 4.29953104, ..., 2.76698155, 2.78100414,
        2.20791581],
       [0.        , 0.        , 0.        , ..., 2.63546884, 4.34269664,
        3.32469143],
       ...,
       [0.        , 0.        , 0.        , ..., 0.        , 2.18534065,
        1.8408748 ],
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        3.09496075],
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        0.        ]])

In [ ]:
Matrix_Dist_Mahalanobis_2(Data=Data_quant) # 22.5 seg

array([[0.        , 2.11289055, 3.7975463 , ..., 4.51559865, 2.31688444,
        1.10588047],
       [0.        , 0.        , 4.35615967, ..., 4.93340427, 2.74011739,
        2.12938584],
       [0.        , 0.        , 0.        , ..., 3.17779509, 3.49432487,
        3.23723317],
       ...,
       [0.        , 0.        , 0.        , ..., 0.        , 3.58695453,
        4.11275247],
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        1.46894947],
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        0.        ]])

In [116]:
def Matrix_Dist_Mahalanobis(Data):

    Data = Data.to_numpy()

    n = len(Data)

    M =  np.empty((n , n))

    S_inv=np.linalg.inv( np.cov(Data , rowvar=False) )

   
    for i in range(0, n):

         for r in range(0, n):

             if i >= r :
               
                 M[i,r] = 0

             else :

                 M[i,r] = Dist_Mahalanobis_3(x = np.array([Data[i,:] - Data[r,:]]) , S_inv=S_inv ) 

                      
    return M 

In [117]:
Matrix_Dist_Mahalanobis(Data=Data_quant.iloc[0:1000, :]) # 4.8 seg

array([[0.        , 2.16321817, 3.81131086, ..., 2.74083537, 3.84799958,
        1.06041471],
       [0.        , 0.        , 4.29953104, ..., 2.76698155, 2.78100414,
        2.20791581],
       [0.        , 0.        , 0.        , ..., 2.63546884, 4.34269664,
        3.32469143],
       ...,
       [0.        , 0.        , 0.        , ..., 0.        , 2.18534065,
        1.8408748 ],
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        3.09496075],
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        0.        ]])

In [ ]:
M_Mahalanobis = Matrix_Dist_Mahalanobis(Data=Data_quant) # 19.7 seg

M_Mahalanobis

array([[0.        , 2.11289055, 3.7975463 , ..., 4.51559865, 2.31688444,
        1.10588047],
       [0.        , 0.        , 4.35615967, ..., 4.93340427, 2.74011739,
        2.12938584],
       [0.        , 0.        , 0.        , ..., 3.17779509, 3.49432487,
        3.23723317],
       ...,
       [0.        , 0.        , 0.        , ..., 0.        , 3.58695453,
        4.11275247],
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        1.46894947],
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        0.        ]])

In [ ]:
M_Mahalanobis = M_Mahalanobis + M_Mahalanobis.T

M_Mahalanobis

array([[0.        , 2.11289055, 3.7975463 , ..., 4.51559865, 2.31688444,
        1.10588047],
       [2.11289055, 0.        , 4.35615967, ..., 4.93340427, 2.74011739,
        2.12938584],
       [3.7975463 , 4.35615967, 0.        , ..., 3.17779509, 3.49432487,
        3.23723317],
       ...,
       [4.51559865, 4.93340427, 3.17779509, ..., 0.        , 3.58695453,
        4.11275247],
       [2.31688444, 2.74011739, 3.49432487, ..., 3.58695453, 0.        ,
        1.46894947],
       [1.10588047, 2.12938584, 3.23723317, ..., 4.11275247, 1.46894947,
        0.        ]])

In [118]:
def a_b_c_d_Matrix(Data):

    X = Data.to_numpy()

    a = X @ X.T

    n = X.shape[0]

    p = X.shape[1]

    ones_matrix = np.ones((n, p)) 

    b = (ones_matrix - X) @ X.T

    c = b.T

    d = (ones_matrix - X) @ (ones_matrix - X).T

    return a , b , c , d , p

In [ ]:
a, b, c, d, p = a_b_c_d_Matrix(Data_binary)

In [ ]:
a

array([[1., 1., 1., ..., 1., 1., 1.],
       [1., 1., 1., ..., 1., 1., 1.],
       [1., 1., 2., ..., 1., 1., 1.],
       ...,
       [1., 1., 1., ..., 1., 1., 1.],
       [1., 1., 1., ..., 1., 2., 1.],
       [1., 1., 1., ..., 1., 1., 1.]])

In [ ]:
Data_binary.iloc[[30,1903], :]

private_garden_recode  private_gym_recode  private_jacuzzi_recode  \
30                      0.0                 1.0                     0.0   
1903                    1.0                 0.0                     0.0   

      balcony_recode  
30               1.0  
1903             1.0

In [ ]:
a[30 , 1903]

1.0

In [ ]:
Data_binary.iloc[[40,1903], :]

private_garden_recode  private_gym_recode  private_jacuzzi_recode  \
40                      0.0                 0.0                     0.0   
1903                    1.0                 0.0                     0.0   

      balcony_recode  
40               1.0  
1903             1.0

In [ ]:
a[40 , 1903]

1.0

In [ ]:
b

array([[0., 0., 1., ..., 0., 1., 0.],
       [0., 0., 1., ..., 0., 1., 0.],
       [0., 0., 0., ..., 0., 1., 0.],
       ...,
       [0., 0., 1., ..., 0., 1., 0.],
       [0., 0., 1., ..., 0., 0., 0.],
       [0., 0., 1., ..., 0., 1., 0.]])

In [ ]:
Data_binary.iloc[[30,1903], :]

private_garden_recode  private_gym_recode  private_jacuzzi_recode  \
30                      0.0                 1.0                     0.0   
1903                    1.0                 0.0                     0.0   

      balcony_recode  
30               1.0  
1903             1.0

In [ ]:
b[30 , 1903]

1.0

In [ ]:
Data_binary.iloc[[40,1903], :]

private_garden_recode  private_gym_recode  private_jacuzzi_recode  \
40                      0.0                 0.0                     0.0   
1903                    1.0                 0.0                     0.0   

      balcony_recode  
40               1.0  
1903             1.0

In [ ]:
b[40,1903]

1.0

In [ ]:
c

array([[0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [1., 1., 0., ..., 1., 1., 1.],
       ...,
       [0., 0., 0., ..., 0., 0., 0.],
       [1., 1., 1., ..., 1., 0., 1.],
       [0., 0., 0., ..., 0., 0., 0.]])

In [ ]:
c[30, 1903]

1.0

In [ ]:
Data_binary.iloc[[40,1903], :]

private_garden_recode  private_gym_recode  private_jacuzzi_recode  \
40                      0.0                 0.0                     0.0   
1903                    1.0                 0.0                     0.0   

      balcony_recode  
40               1.0  
1903             1.0

In [ ]:
c[40, 1903]

0.0

In [ ]:
d

array([[3., 3., 2., ..., 3., 2., 3.],
       [3., 3., 2., ..., 3., 2., 3.],
       [2., 2., 2., ..., 2., 1., 2.],
       ...,
       [3., 3., 2., ..., 3., 2., 3.],
       [2., 2., 1., ..., 2., 2., 2.],
       [3., 3., 2., ..., 3., 2., 3.]])

In [ ]:
d[30,1903]

1.0

In [ ]:
Data_binary.iloc[[40,1903], :]

private_garden_recode  private_gym_recode  private_jacuzzi_recode  \
40                      0.0                 0.0                     0.0   
1903                    1.0                 0.0                     0.0   

      balcony_recode  
40               1.0  
1903             1.0

In [ ]:
d[40,1903]

2.0

## Similaridad de Sokal

In [119]:
def Sokal_Similarity_1(i , r, Data):

    a, b, c, d, p = a_b_c_d_Matrix(Data)

    if a[i,r] + d[i,r] == 0 :

        Sokal_Similarity = 0

        
    else :

        Sokal_Similarity = (a[i,r] + d[i,r]) / p
        

    return Sokal_Similarity

In [ ]:
Sokal_Similarity_1(i=1 , r=2, Data=Data_binary)

0.75

In [120]:
def Sokal_Similarity_2(i , r, a , d, p):


    if a[i,r] + d[i,r] == 0 :

        Sokal_Similarity = 0

        
    else :

        Sokal_Similarity = (a[i,r] + d[i,r]) / p
        

    return Sokal_Similarity

In [ ]:
a, b, c, d, p = a_b_c_d_Matrix(Data_binary)

Sokal_Similarity_2(i=1, r=2, a=a, d=d, p=p)

0.75

In [121]:
def Matrix_Sim_Sokal_1(Data):

    n = len(Data)

    M =  np.empty((n , n))

   
    for i in range(0, n):

         for r in range(0, n):

            if i > r :
               
                 M[i,r] = 0
            
            elif i == r :
               
                 M[i,r] = 1
            
            else :

                 M[i,r] = Sokal_Similarity_1(i, r, Data)
                      
    return M 

In [ ]:
Matrix_Sim_Sokal_1(Data=Data_binary.iloc[0:300 , 0:300]) # 1.29 mins

array([[1.  , 1.  , 0.75, ..., 0.75, 0.75, 1.  ],
       [0.  , 1.  , 0.75, ..., 0.75, 0.75, 1.  ],
       [0.  , 0.  , 1.  , ..., 0.5 , 0.5 , 0.75],
       ...,
       [0.  , 0.  , 0.  , ..., 1.  , 1.  , 0.75],
       [0.  , 0.  , 0.  , ..., 0.  , 1.  , 0.75],
       [0.  , 0.  , 0.  , ..., 0.  , 0.  , 1.  ]])

In [122]:
def Matrix_Sim_Sokal_2(Data):

    n = len(Data)

    M =  np.empty((n , n))

    a, b, c, d, p = a_b_c_d_Matrix(Data)

   
    for i in range(0, n):

         for r in range(0, n):

            if i > r :
               
                 M[i,r] = 0
            
            elif i == r :
               
                 M[i,r] = 1

            else :

                 M[i,r] = Sokal_Similarity_2(i=i, r=r, a=a, d=d, p=p)
                      
    return M 

In [ ]:
Matrix_Sim_Sokal_2(Data=Data_binary.iloc[0:500 , 0:500]) # 0.2 seg

array([[1.  , 1.  , 0.75, ..., 1.  , 1.  , 0.75],
       [0.  , 1.  , 0.75, ..., 1.  , 1.  , 0.75],
       [0.  , 0.  , 1.  , ..., 0.75, 0.75, 0.5 ],
       ...,
       [0.  , 0.  , 0.  , ..., 1.  , 1.  , 0.75],
       [0.  , 0.  , 0.  , ..., 0.  , 1.  , 0.75],
       [0.  , 0.  , 0.  , ..., 0.  , 0.  , 1.  ]])

In [ ]:
M_Sim_Sokal = Matrix_Sim_Sokal_2(Data=Data_binary) # 3.4 seg

M_Sim_Sokal

array([[1.  , 1.  , 0.75, ..., 1.  , 0.75, 1.  ],
       [0.  , 1.  , 0.75, ..., 1.  , 0.75, 1.  ],
       [0.  , 0.  , 1.  , ..., 0.75, 0.5 , 0.75],
       ...,
       [0.  , 0.  , 0.  , ..., 1.  , 0.75, 1.  ],
       [0.  , 0.  , 0.  , ..., 0.  , 1.  , 0.75],
       [0.  , 0.  , 0.  , ..., 0.  , 0.  , 1.  ]])

In [ ]:
M_Sim_Sokal = M_Sim_Sokal + M_Sim_Sokal.T - np.diag(np.repeat(1 , len(M_Sim_Sokal)), k=0)

M_Sim_Sokal

array([[1.  , 1.  , 0.75, ..., 1.  , 0.75, 1.  ],
       [1.  , 1.  , 0.75, ..., 1.  , 0.75, 1.  ],
       [0.75, 0.75, 1.  , ..., 0.75, 0.5 , 0.75],
       ...,
       [1.  , 1.  , 0.75, ..., 1.  , 0.75, 1.  ],
       [0.75, 0.75, 0.5 , ..., 0.75, 1.  , 0.75],
       [1.  , 1.  , 0.75, ..., 1.  , 0.75, 1.  ]])

In [123]:
def Dist_Sokal(i, r, a, d, p):

    dist_Sokal = np.sqrt( Sokal_Similarity_2(i , i, a , d, p) + Sokal_Similarity_2(r , r, a , d, p) - 2*Sokal_Similarity_2(i , r, a , d, p) )

    return dist_Sokal  

In [ ]:
a, b, c, d, p = a_b_c_d_Matrix(Data_binary)

Dist_Sokal(i=1, r=2, a=a, d=d, p=p)

0.7071067811865476

In [124]:
def Matrix_Dist_Sokal(Data):

    n = len(Data)

    M =  np.empty((n , n))

    a, b, c, d, p = a_b_c_d_Matrix(Data)

   
    for i in range(0, n):

         for r in range(0, n):

             if i >= r :
               
                 M[i,r] = 0

             else :

                 M[i,r] = Dist_Sokal(i=i, r=r, a=a, d=d, p=p)
                      
    return M 

In [125]:
M_Dist_Sokal = Matrix_Dist_Sokal(Data=Data_binary)

M_Dist_Sokal

array([[0.        , 0.        , 0.70710678, ..., 0.        , 0.70710678,
        0.        ],
       [0.        , 0.        , 0.70710678, ..., 0.        , 0.70710678,
        0.        ],
       [0.        , 0.        , 0.        , ..., 0.70710678, 1.        ,
        0.70710678],
       ...,
       [0.        , 0.        , 0.        , ..., 0.        , 0.70710678,
        0.        ],
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        0.70710678],
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        0.        ]])

In [ ]:
M_Dist_Sokal = M_Dist_Sokal + M_Dist_Sokal.T

M_Dist_Sokal

array([[0.        , 0.        , 0.70710678, ..., 0.        , 0.70710678,
        0.        ],
       [0.        , 0.        , 0.70710678, ..., 0.        , 0.70710678,
        0.        ],
       [0.70710678, 0.70710678, 0.        , ..., 0.70710678, 1.        ,
        0.70710678],
       ...,
       [0.        , 0.        , 0.70710678, ..., 0.        , 0.70710678,
        0.        ],
       [0.70710678, 0.70710678, 1.        , ..., 0.70710678, 0.        ,
        0.70710678],
       [0.        , 0.        , 0.70710678, ..., 0.        , 0.70710678,
        0.        ]])

## Similaridad de Jaccard

In [126]:
def Jaccard_Similarity_1(i, r, Data):

    a, b, c, d, p = a_b_c_d_Matrix(Data)
 

    if a[i,r] == 0 and  (p-d[i,r]) == 0 :

        Jaccard_Similarity = 1

    else :

        Jaccard_Similarity = a[i,r] / (p-d[i,r])


    return Jaccard_Similarity

In [ ]:
Jaccard_Similarity_1(i=2, r=5, Data=Data_binary)

0.0

In [127]:
def Jaccard_Similarity_2(i , r, a , b, c):


    if a[i,r] == 0 and  (p-d[i,r]) == 0 :

        Jaccard_Similarity = 1

    else :

        Jaccard_Similarity = a[i,r] / (p-d[i,r])
        

    return Jaccard_Similarity

In [ ]:
a, b, c, d, p = a_b_c_d_Matrix(Data=Data_binary)

Jaccard_Similarity_2(i=2, r=5, a=a , b=b, c=c)

0.0

In [128]:
def Matrix_Sim_Jaccard_1(Data):

    n = len(Data)

    M =  np.empty((n , n))

   
    for i in range(0, n):

         for r in range(0, n):

            if i > r :
               
                 M[i,r] = 0
            
            elif i == r :
               
                 M[i,r] = 1
            
            else :

                 M[i,r] = Jaccard_Similarity_1(i, r, Data)
                      
    return M 

In [ ]:
Matrix_Sim_Jaccard_1(Data=Data_binary.iloc[0:300, 0:300])  # 1.52 min

array([[1. , 1. , 0.5, ..., 0. , 0. , 1. ],
       [0. , 1. , 0.5, ..., 0. , 0. , 1. ],
       [0. , 0. , 1. , ..., 0. , 0. , 0.5],
       ...,
       [0. , 0. , 0. , ..., 1. , 1. , 0. ],
       [0. , 0. , 0. , ..., 0. , 1. , 0. ],
       [0. , 0. , 0. , ..., 0. , 0. , 1. ]])

In [129]:
def Matrix_Sim_Jaccard_2(Data):

    n = len(Data)

    M =  np.empty((n , n))

    a, b, c, d, p = a_b_c_d_Matrix(Data=Data)

   
    for i in range(0, n):

         for r in range(0, n):

            if i > r :
               
                 M[i,r] = 0
            
            elif i == r :
               
                 M[i,r] = 1
            
            else :

                 M[i,r] = Jaccard_Similarity_2(i , r , a , b , c)
                      
    return M 

In [ ]:
Matrix_Sim_Jaccard_2(Data=Data_binary.iloc[0:300,0:300])

array([[1. , 1. , 0.5, ..., 0. , 0. , 1. ],
       [0. , 1. , 0.5, ..., 0. , 0. , 1. ],
       [0. , 0. , 1. , ..., 0. , 0. , 0.5],
       ...,
       [0. , 0. , 0. , ..., 1. , 1. , 0. ],
       [0. , 0. , 0. , ..., 0. , 1. , 0. ],
       [0. , 0. , 0. , ..., 0. , 0. , 1. ]])

In [ ]:
M_Jaccard = Matrix_Sim_Jaccard_2(Data=Data_binary)

M_Jaccard 

array([[1.        , 1.        , 0.5       , ..., 1.        , 0.5       ,
        1.        ],
       [0.        , 1.        , 0.5       , ..., 1.        , 0.5       ,
        1.        ],
       [0.        , 0.        , 1.        , ..., 0.5       , 0.33333333,
        0.5       ],
       ...,
       [0.        , 0.        , 0.        , ..., 1.        , 0.5       ,
        1.        ],
       [0.        , 0.        , 0.        , ..., 0.        , 1.        ,
        0.5       ],
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        1.        ]])

In [ ]:
M_Jaccard = M_Jaccard + M_Jaccard.T - np.diag(np.repeat(1 , len(M_Jaccard)), k=0)

M_Jaccard 

array([[1.        , 1.        , 0.5       , ..., 1.        , 0.5       ,
        1.        ],
       [1.        , 1.        , 0.5       , ..., 1.        , 0.5       ,
        1.        ],
       [0.5       , 0.5       , 1.        , ..., 0.5       , 0.33333333,
        0.5       ],
       ...,
       [1.        , 1.        , 0.5       , ..., 1.        , 0.5       ,
        1.        ],
       [0.5       , 0.5       , 0.33333333, ..., 0.5       , 1.        ,
        0.5       ],
       [1.        , 1.        , 0.5       , ..., 1.        , 0.5       ,
        1.        ]])

In [130]:
def Dist_Jaccard(i , r, a , b, c):

    Dist_Jaccard = np.sqrt( Jaccard_Similarity_2(i , i, a , b, c) + Jaccard_Similarity_2(r , r, a , b, c) - 2*Jaccard_Similarity_2(i , r, a , b, c) )

    return Dist_Jaccard  

In [ ]:
Data_binary.iloc[[2,5],:]

private_garden_recode  private_gym_recode  private_jacuzzi_recode  \
2                    0.0                 0.0                     1.0   
5                    0.0                 0.0                     0.0   

   balcony_recode  
2             1.0  
5             0.0

In [ ]:
a, b, c, d, p = a_b_c_d_Matrix(Data_binary)

Dist_Jaccard(i=2 , r=5, a=a , b=b, c=c)

1.4142135623730951

In [ ]:
Jaccard_Similarity_2(2 , 2, a , b, c) 

1.0

In [ ]:
Jaccard_Similarity_2(5 , 5, a , b, c) 

1

In [ ]:
2*Jaccard_Similarity_2(2 , 5, a , b, c)

0.0

In [ ]:
np.sqrt(Jaccard_Similarity_2(2 , 2, a , b, c) + Jaccard_Similarity_2(5 , 5, a , b, c)  -2*Jaccard_Similarity_2(2 , 5, a , b, c) )

1.4142135623730951

In [131]:
def Matrix_Dist_Jaccard(Data):

    n = len(Data)

    M =  np.empty((n , n))

    a, b, c, d, p = a_b_c_d_Matrix(Data)

   
    for i in range(0, n):

         for r in range(0, n):

             if i >= r :
               
                 M[i,r] = 0

             else :

                 M[i,r] = Dist_Jaccard(i=i, r=r, a=a, b=b, c=c)
                      
    return M 

In [ ]:
M_Dist_Jaccard = Matrix_Dist_Jaccard(Data=Data_binary)

M_Dist_Jaccard

array([[0.        , 0.        , 1.        , ..., 0.        , 1.        ,
        0.        ],
       [0.        , 0.        , 1.        , ..., 0.        , 1.        ,
        0.        ],
       [0.        , 0.        , 0.        , ..., 1.        , 1.15470054,
        1.        ],
       ...,
       [0.        , 0.        , 0.        , ..., 0.        , 1.        ,
        0.        ],
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        1.        ],
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        0.        ]])

In [ ]:
M_Dist_Jaccard = M_Dist_Jaccard + M_Dist_Jaccard.T 

M_Dist_Jaccard

array([[0.        , 0.        , 1.        , ..., 0.        , 1.        ,
        0.        ],
       [0.        , 0.        , 1.        , ..., 0.        , 1.        ,
        0.        ],
       [1.        , 1.        , 0.        , ..., 1.        , 1.15470054,
        1.        ],
       ...,
       [0.        , 0.        , 1.        , ..., 0.        , 1.        ,
        0.        ],
       [1.        , 1.        , 1.15470054, ..., 1.        , 0.        ,
        1.        ],
       [0.        , 0.        , 1.        , ..., 0.        , 1.        ,
        0.        ]])

## Similaridad de coincidencias

In [132]:
def alpha(x_i, x_r):

    alpha = sum(x_i == x_r)

    return(alpha)   

In [ ]:
Data_multiclass.iloc[2,:] 

quality_recode    2.0
rating            5.0
type              1.0
Name: 2, dtype: float64

In [ ]:
Data_multiclass.iloc[5,:] 

quality_recode    2.0
rating            5.0
type              1.0
Name: 5, dtype: float64

In [ ]:
alpha(x_i=Data_multiclass.iloc[2,:] , x_r=Data_multiclass.iloc[5,:])

3

In [ ]:
Data_multiclass.iloc[15,:] 

quality_recode    2.0
rating            3.0
type              2.0
Name: 15, dtype: float64

In [ ]:
Data_multiclass.iloc[14,:]

quality_recode    2.0
rating            4.0
type              1.0
Name: 14, dtype: float64

In [ ]:
alpha(x_i=Data_multiclass.iloc[14,:] , x_r=Data_multiclass.iloc[15,:])

1

In [133]:
def matching_similarity(x_i, x_r, Data):

    p = Data.shape[1]

    matches_similarity = alpha(x_i, x_r) / p

    return(matches_similarity)

In [ ]:
matching_similarity(x_i=Data_multiclass.iloc[2,:] , x_r=Data_multiclass.iloc[5,:], Data=Data_multiclass)

1.0

In [ ]:
matching_similarity(x_i=Data_multiclass.iloc[14,:] , x_r=Data_multiclass.iloc[15,:], Data=Data_multiclass)

0.3333333333333333

In [ ]:
def Matrix_Sim_Matching(Data):
    
    Data = Data.to_numpy()

    n = len(Data)

    M =  np.empty((n , n))

   
    for i in range(0, n):

         for r in range(0, n):

            if i > r :
               
                 M[i,r] = 0
            
            elif i == r :
               
                 M[i,r] = 1
            
            else :

                 M[i,r] = matches_similarity(x_i=Data[i,:], x_r=Data[r,:], Data=Data)
                      
    return M 

In [ ]:
M_matches = Matrix_Sim_Matching(Data=Data_multiclass)

M_matches 

array([[1.        , 0.33333333, 0.33333333, ..., 0.33333333, 0.33333333,
        0.33333333],
       [0.        , 1.        , 0.66666667, ..., 1.        , 0.66666667,
        0.        ],
       [0.        , 0.        , 1.        , ..., 0.66666667, 0.66666667,
        0.        ],
       ...,
       [0.        , 0.        , 0.        , ..., 1.        , 0.66666667,
        0.        ],
       [0.        , 0.        , 0.        , ..., 0.        , 1.        ,
        0.        ],
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        1.        ]])

In [ ]:
M_matches = M_matches + M_matches.T - np.diag(np.repeat(1 , len(M_matches)), k=0)

M_matches

array([[1.        , 0.33333333, 0.33333333, ..., 0.33333333, 0.33333333,
        0.33333333],
       [0.33333333, 1.        , 0.66666667, ..., 1.        , 0.66666667,
        0.        ],
       [0.33333333, 0.66666667, 1.        , ..., 0.66666667, 0.66666667,
        0.        ],
       ...,
       [0.33333333, 1.        , 0.66666667, ..., 1.        , 0.66666667,
        0.        ],
       [0.33333333, 0.66666667, 0.66666667, ..., 0.66666667, 1.        ,
        0.        ],
       [0.33333333, 0.        , 0.        , ..., 0.        , 0.        ,
        1.        ]])

In [145]:
def Dist_Matching(x_i, x_r, Data):

    Dist_Matches = np.sqrt(matches_similarity(x_i, x_i, Data) +  matches_similarity(x_r, x_r, Data) - 2*matches_similarity(x_i, x_r, Data) )

    return( Dist_Matches )

In [143]:
def Matrix_Dist_Matching(Data):

    Data = Data.to_numpy()

    n = len(Data)

    M =  np.empty((n , n))
   
    for i in range(0, n):

         for r in range(0, n):

             if i >= r :
               
                M[i,r] = 0

             else :

                M[i,r] = Dist_Matching(x_i=Data[i,:], x_r=Data[r,:], Data=Data)
                      
    return M 

In [ ]:
M_Dist_Matching = Matrix_Dist_Matching(Data=Data_multiclass)

M_Dist_Matching

array([[0.        , 1.15470054, 1.15470054, ..., 1.15470054, 1.15470054,
        1.15470054],
       [0.        , 0.        , 0.81649658, ..., 0.        , 0.81649658,
        1.41421356],
       [0.        , 0.        , 0.        , ..., 0.81649658, 0.81649658,
        1.41421356],
       ...,
       [0.        , 0.        , 0.        , ..., 0.        , 0.81649658,
        1.41421356],
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        1.41421356],
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        0.        ]])

In [ ]:
M_Dist_Matching = M_Dist_Matching + M_Dist_Matching.T

M_Dist_Matching

array([[0.        , 1.15470054, 1.15470054, ..., 1.15470054, 1.15470054,
        1.15470054],
       [1.15470054, 0.        , 0.81649658, ..., 0.        , 0.81649658,
        1.41421356],
       [1.15470054, 0.81649658, 0.        , ..., 0.81649658, 0.81649658,
        1.41421356],
       ...,
       [1.15470054, 0.        , 0.81649658, ..., 0.        , 0.81649658,
        1.41421356],
       [1.15470054, 0.81649658, 0.81649658, ..., 0.81649658, 0.        ,
        1.41421356],
       [1.15470054, 1.41421356, 1.41421356, ..., 1.41421356, 1.41421356,
        0.        ]])

## Similaridad de Gower

In [134]:
def Matrix_Gower_Similarity(Data, p1, p2, p3 ):

    # Data tiene que ser un numpy array tal que:
    # las primeras p1 tiene que ser variables cuantitativas, 
    # las siguientes p2 binarias y las restantes p3 multicales

    # Si Data no contiene variables cuantitativas --> p1=0
    # Si Data no contiene variables binarias --> p2=0
    # Si Data no contiene variables multiclase --> p3=0

    Data_pd = Data

    Data = Data.to_numpy()

    n = len(Data)

    M =  np.empty((n , n))

############################################################
    
    G_vector = np.repeat(0.5, p1)

    for k in range(0, p1):

        G_vector[k] = Data[:,k].max() - Data[:,k].min()

############################################################
   
    ones = np.repeat(1, p1)

    Quant_Data = Data[: , 0:p1]

    Binary_Data_pd = Data_pd.iloc[: , (p1):(p1+p2)]

    Multiple_Data = Data[: , (p1+p2):(p1+p2+p3) ]

############################################################

    a, b, c, d, p = a_b_c_d_Matrix(Binary_Data_pd)

############################################################

    for i in range(0, n):

         for r in range(0, n):

            if i > r :
               
                 M[i,r] = 0
            
            elif i == r :
               
                 M[i,r] = 1
            
            else :

                # Calculo de la distancia de Gower entre las observaciones x_i y x_r

                numerator_part_1 = ( ones - ( abs(Quant_Data[i,:] - Quant_Data[r,:]) / G_vector ) ).sum() 

                numerator_part_2 = a[i,r] 
                
                numerator_part_3 = alpha(Multiple_Data[i,:], Multiple_Data[r,:])

                numerator = numerator_part_1 + numerator_part_2 + numerator_part_3

                denominator = p1 + (p2 - d[i,r]) + p3
                

                if denominator == 0:

                    M[i,r] = 0

                else:    
                    
                    M[i,r] = numerator / denominator
                      
   
    return M  

In [ ]:
Data_mixed.head()

latitude  longitude    price  size_in_m_2  no_of_bedrooms  \
0  25.113208  55.138932  2700000   100.242337               1   
1  25.106809  55.151201  2850000   146.972546               2   
2  25.063302  55.137728  1150000   181.253753               3   
3  25.227295  55.341761  2850000   187.664060               2   
4  25.114275  55.139764  1729200    47.101821               0   

   no_of_bathrooms  private_garden_recode  private_gym_recode  \
0                2                    0.0                 0.0   
1                2                    0.0                 0.0   
2                5                    0.0                 0.0   
3                3                    0.0                 0.0   
4                1                    0.0                 0.0   

   private_jacuzzi_recode  balcony_recode  quality_recode  rating  type  
0                     0.0             1.0             2.0     3.0   2.0  
1                     0.0             1.0             2.0     2.0   1.0  
2                     1.0             1.0             2.0     5.0   1.0  
3                     0.0             1.0             1.0     5.0   2.0  
4                     0.0             0.0             2.0     1.0   2.0

In [135]:
M_Gower = Matrix_Gower_Similarity(Data=Data_mixed, p1=6, p2=4, p3=3)

M_Gower

array([[1.        , 0.76928447, 0.61234779, ..., 0.61200705, 0.68768535,
        0.77882015],
       [0.        , 1.        , 0.72440439, ..., 0.83958297, 0.75951698,
        0.65783494],
       [0.        , 0.        , 1.        , ..., 0.73084198, 0.63188025,
        0.52548719],
       ...,
       [0.        , 0.        , 0.        , ..., 1.        , 0.65208854,
        0.4993045 ],
       [0.        , 0.        , 0.        , ..., 0.        , 1.        ,
        0.61603067],
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        1.        ]])

In [ ]:
M_Gower = M_Gower  + M_Gower.T - np.diag(np.repeat(1 , len(M_Gower)), k=0)

M_Gower

array([[1.        , 0.76928447, 0.61234779, ..., 0.61200705, 0.68768535,
        0.77882015],
       [0.76928447, 1.        , 0.72440439, ..., 0.83958297, 0.75951698,
        0.65783494],
       [0.61234779, 0.72440439, 1.        , ..., 0.73084198, 0.63188025,
        0.52548719],
       ...,
       [0.61200705, 0.83958297, 0.73084198, ..., 1.        , 0.65208854,
        0.4993045 ],
       [0.68768535, 0.75951698, 0.63188025, ..., 0.65208854, 1.        ,
        0.61603067],
       [0.77882015, 0.65783494, 0.52548719, ..., 0.4993045 , 0.61603067,
        1.        ]])

In [137]:
def Matrix_Dist_Gower(Data, p1, p2, p3 ):

    M = Matrix_Gower_Similarity(Data, p1, p2, p3)

    M = M + M.T - np.diag(np.repeat(1 , len(M)), k=0)

    M = np.sqrt( 1 - M )

    return M

In [138]:
M_Gower_Dist = Matrix_Dist_Gower(Data=Data_mixed, p1=6, p2=4, p3=3)

M_Gower_Dist

array([[0.        , 0.48032856, 0.62261723, ..., 0.6228908 , 0.55885118,
        0.47029762],
       [0.48032856, 0.        , 0.52497201, ..., 0.40052095, 0.49039069,
        0.58494877],
       [0.62261723, 0.52497201, 0.        , ..., 0.51880441, 0.60672873,
        0.6888489 ],
       ...,
       [0.6228908 , 0.40052095, 0.51880441, ..., 0.        , 0.5898402 ,
        0.70759841],
       [0.55885118, 0.49039069, 0.60672873, ..., 0.5898402 , 0.        ,
        0.61965259],
       [0.47029762, 0.58494877, 0.6888489 , ..., 0.70759841, 0.61965259,
        0.        ]])

### Probando la similaridad de Gower con conjuntos de datos no necesariamente cuantitativos-binarios-multiclase:

In [ ]:
Matrix_Gower_Similarity(Data=Data_quant, p1=6, p2=0, p3=0)

array([[1.        , 0.94880745, 0.78930428, ..., 0.68667841, 0.92742315,
        0.96470026],
       [0.        , 1.        , 0.82807471, ..., 0.73263828, 0.89244779,
        0.92972489],
       [0.        , 0.        , 1.        , ..., 0.83987697, 0.76376051,
        0.79672652],
       ...,
       [0.        , 0.        , 0.        , ..., 1.        , 0.69549566,
        0.6655075 ],
       [0.        , 0.        , 0.        , ..., 0.        , 1.        ,
        0.96272289],
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        1.        ]])

In [ ]:
Matrix_Gower_Similarity(Data=Data_binary, p1=0, p2=4, p3=0)

# Coincide con la matriz de similaridades de Jaccard que se ha obtenido anteriormente.

array([[1.        , 1.        , 0.5       , ..., 1.        , 0.5       ,
        1.        ],
       [0.        , 1.        , 0.5       , ..., 1.        , 0.5       ,
        1.        ],
       [0.        , 0.        , 1.        , ..., 0.5       , 0.33333333,
        0.5       ],
       ...,
       [0.        , 0.        , 0.        , ..., 1.        , 0.5       ,
        1.        ],
       [0.        , 0.        , 0.        , ..., 0.        , 1.        ,
        0.5       ],
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        1.        ]])

In [ ]:
Matrix_Gower_Similarity(Data=Data_multiclass, p1=0, p2=0, p3=3)

# Coincide con la matriz de similaridades por coincidencias que se ha obtenido anteriormente.

array([[1.        , 0.33333333, 0.33333333, ..., 0.33333333, 0.33333333,
        0.33333333],
       [0.        , 1.        , 0.66666667, ..., 1.        , 0.66666667,
        0.        ],
       [0.        , 0.        , 1.        , ..., 0.66666667, 0.66666667,
        0.        ],
       ...,
       [0.        , 0.        , 0.        , ..., 1.        , 0.66666667,
        0.        ],
       [0.        , 0.        , 0.        , ..., 0.        , 1.        ,
        0.        ],
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        1.        ]])

In [ ]:
Data_binary_multiclass = pd.concat([Data_binary,Data_multiclass], axis=1)

In [ ]:
Matrix_Gower_Similarity(Data=Data_binary_multiclass , p1=0, p2=4, p3=3)

array([[1.  , 0.5 , 0.4 , ..., 0.5 , 0.4 , 0.5 ],
       [0.  , 1.  , 0.6 , ..., 1.  , 0.6 , 0.25],
       [0.  , 0.  , 1.  , ..., 0.6 , 0.5 , 0.2 ],
       ...,
       [0.  , 0.  , 0.  , ..., 1.  , 0.6 , 0.25],
       [0.  , 0.  , 0.  , ..., 0.  , 1.  , 0.2 ],
       [0.  , 0.  , 0.  , ..., 0.  , 0.  , 1.  ]])

In [ ]:
Data_quant_multiclass = pd.concat([Data_quant,Data_multiclass], axis=1)

In [ ]:
Matrix_Gower_Similarity(Data=Data_quant_multiclass, p1=6, p2=0, p3=3)

array([[1.        , 0.74364941, 0.63731397, ..., 0.56889672, 0.72939321,
        0.75424461],
       [0.        , 1.        , 0.77427203, ..., 0.82175886, 0.81718741,
        0.6198166 ],
       [0.        , 0.        , 1.        , ..., 0.7821402 , 0.73139589,
        0.53115102],
       ...,
       [0.        , 0.        , 0.        , ..., 1.        , 0.68588599,
        0.44367166],
       [0.        , 0.        , 0.        , ..., 0.        , 1.        ,
        0.64181526],
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        1.        ]])

In [ ]:
Data_quant_binary = pd.concat([Data_quant,Data_binary], axis=1)

In [ ]:
Matrix_Gower_Similarity(Data=Data_quant_binary , p1=6, p2=4, p3=0)

array([[1.        , 0.95612067, 0.71697821, ..., 0.73143864, 0.82056736,
        0.96974308],
       [0.        , 1.        , 0.74605603, ..., 0.77083282, 0.79433584,
        0.9397642 ],
       [0.        , 0.        , 1.        , ..., 0.75490772, 0.62028478,
        0.72254489],
       ...,
       [0.        , 0.        , 0.        , ..., 1.        , 0.64662174,
        0.71329214],
       [0.        , 0.        , 0.        , ..., 0.        , 1.        ,
        0.84704217],
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        1.        ]])

### Comparacion con los resultados obtenidos con la funcion Gower2.m de Aurea  --> se obtienen los mismos resultados !!

In [ ]:
futbol = pd.read_csv('jugadores_futbol.csv')

futbol_np = futbol.to_numpy()

In [ ]:
M_Gower_futbol = Matrix_Gower_Similarity(Data=futbol_np, p1=4, p2=1, p3=2)

In [ ]:
M_Gower_futbol = M_Gower_futbol  + M_Gower_futbol.T - np.diag(np.repeat(1 , len(M_Gower_futbol)), k=0)

M_Gower_futbol

array([[1.        , 0.63275613, 0.45075758, ..., 0.41684704, 0.61201299,
        0.39303752],
       [0.63275613, 1.        , 0.43076599, ..., 0.49810606, 0.64920034,
        0.47032828],
       [0.45075758, 0.43076599, 1.        , ..., 0.63720539, 0.7260101 ,
        0.60942761],
       ...,
       [0.41684704, 0.49810606, 0.63720539, ..., 1.        , 0.67508418,
        0.91666667],
       [0.61201299, 0.64920034, 0.7260101 , ..., 0.67508418, 1.        ,
        0.6473064 ],
       [0.39303752, 0.47032828, 0.60942761, ..., 0.91666667, 0.6473064 ,
        1.        ]])

In [ ]:
M_Gower_futbol[0:8, 0:8]

array([[1.        , 0.63275613, 0.45075758, 0.38870851, 0.42676768,
        0.61471861, 0.56349206, 0.42839105],
       [0.63275613, 1.        , 0.43076599, 0.37689394, 0.46338384,
        0.44660895, 0.5256734 , 0.43244949],
       [0.45075758, 0.43076599, 1.        , 0.58038721, 0.64877946,
        0.58928571, 0.34995791, 0.52946128],
       [0.38870851, 0.37689394, 0.58038721, 1.        , 0.5625    ,
        0.52723665, 0.27756734, 0.625     ],
       [0.42676768, 0.46338384, 0.64877946, 0.5625    , 1.        ,
        0.77236652, 0.46506734, 0.6875    ],
       [0.61471861, 0.44660895, 0.58928571, 0.52723665, 0.77236652,
        1.        , 0.42496392, 0.63834776],
       [0.56349206, 0.5256734 , 0.34995791, 0.27756734, 0.46506734,
        0.42496392, 1.        , 0.42571549],
       [0.42839105, 0.43244949, 0.52946128, 0.625     , 0.6875    ,
        0.63834776, 0.42571549, 1.        ]])

In [ ]:
M_Gower_Dist_futbol = Matrix_Gower_Distance(Data=futbol_np, p1=4, p2=1, p3=2)

M_Gower_Dist_futbol

array([[0.        , 0.60600649, 0.74110892, ..., 0.76364452, 0.62288604,
        0.77907797],
       [0.60600649, 0.        , 0.75447598, ..., 0.70844473, 0.59228343,
        0.72778549],
       [0.74110892, 0.75447598, 0.        , ..., 0.60232434, 0.52344044,
        0.62495791],
       ...,
       [0.76364452, 0.70844473, 0.60232434, ..., 0.        , 0.57001388,
        0.28867513],
       [0.62288604, 0.59228343, 0.52344044, ..., 0.57001388, 0.        ,
        0.59388012],
       [0.77907797, 0.72778549, 0.62495791, ..., 0.28867513, 0.59388012,
        0.        ]])

In [ ]:
M_Gower_Dist_futbol[0:8 , 0:8]**2

array([[0.        , 0.36724387, 0.54924242, 0.61129149, 0.57323232,
        0.38528139, 0.43650794, 0.57160895],
       [0.36724387, 0.        , 0.56923401, 0.62310606, 0.53661616,
        0.55339105, 0.4743266 , 0.56755051],
       [0.54924242, 0.56923401, 0.        , 0.41961279, 0.35122054,
        0.41071429, 0.65004209, 0.47053872],
       [0.61129149, 0.62310606, 0.41961279, 0.        , 0.4375    ,
        0.47276335, 0.72243266, 0.375     ],
       [0.57323232, 0.53661616, 0.35122054, 0.4375    , 0.        ,
        0.22763348, 0.53493266, 0.3125    ],
       [0.38528139, 0.55339105, 0.41071429, 0.47276335, 0.22763348,
        0.        , 0.57503608, 0.36165224],
       [0.43650794, 0.4743266 , 0.65004209, 0.72243266, 0.53493266,
        0.57503608, 0.        , 0.57428451],
       [0.57160895, 0.56755051, 0.47053872, 0.375     , 0.3125    ,
        0.36165224, 0.57428451, 0.        ]])

 Primeras 8 filas y 8 columnas obtenidas con la funcion Gower2.m de Aurea para la matriz de cuadrados de distancias de Gower
  
         0    0.3672    0.5492    0.6113    0.5732    0.3853    0.4365    0.5716
    0.3672         0    0.5692    0.6231    0.5366    0.5534    0.4743    0.5676
    0.5492    0.5692         0    0.4196    0.3512    0.4107    0.6500    0.4705
    0.6113    0.6231    0.4196         0    0.4375    0.4728    0.7224    0.3750
    0.5732    0.5366    0.3512    0.4375         0    0.2276    0.5349    0.3125
    0.3853    0.5534    0.4107    0.4728    0.2276         0    0.5750    0.3617
    0.4365    0.4743    0.6500    0.7224    0.5349    0.5750         0    0.5743
    0.5716    0.5676    0.4705    0.3750    0.3125    0.3617    0.5743         0



### Comparacion con lo que se obtiene al usar el paquete Gower  --> No se obtienen los mismos resultados.

In [ ]:
import gower

M_Dist_cuadrados_Gower_futbol_Paquete = gower.gower_matrix(futbol_np, cat_features=[False,False,False,False,True,True,True])

M_Dist_cuadrados_Gower_futbol_Paquete[0:8 , 0:8]

array([[0.        , 0.36724386, 0.54924244, 0.61129147, 0.57323235,
        0.38528138, 0.43650794, 0.57160896],
       [0.36724386, 0.        , 0.48791486, 0.53409094, 0.4599567 ,
        0.55339104, 0.40656567, 0.48647186],
       [0.54924244, 0.48791486, 0.        , 0.3596681 , 0.3010462 ,
        0.4107143 , 0.5571789 , 0.4033189 ],
       [0.61129147, 0.53409094, 0.3596681 , 0.        , 0.375     ,
        0.47276336, 0.619228  , 0.32142857],
       [0.57323235, 0.4599567 , 0.3010462 , 0.375     , 0.        ,
        0.22763348, 0.4585137 , 0.26785713],
       [0.38528138, 0.55339104, 0.4107143 , 0.47276336, 0.22763348,
        0.        , 0.5750361 , 0.36165223],
       [0.43650794, 0.40656567, 0.5571789 , 0.619228  , 0.4585137 ,
        0.5750361 , 0.        , 0.4922439 ],
       [0.57160896, 0.48647186, 0.4033189 , 0.32142857, 0.26785713,
        0.36165223, 0.4922439 , 0.        ]], dtype=float32)

 Primeras 8 filas y 8 columnas obtenidas con la funcion Gower2.m de Aurea para la matriz de cuadrados de distancias de Gower
  
         0    0.3672    0.5492    0.6113    0.5732    0.3853    0.4365    0.5716
    0.3672         0    0.5692    0.6231    0.5366    0.5534    0.4743    0.5676
    0.5492    0.5692         0    0.4196    0.3512    0.4107    0.6500    0.4705
    0.6113    0.6231    0.4196         0    0.4375    0.4728    0.7224    0.3750
    0.5732    0.5366    0.3512    0.4375         0    0.2276    0.5349    0.3125
    0.3853    0.5534    0.4107    0.4728    0.2276         0    0.5750    0.3617
    0.4365    0.4743    0.6500    0.7224    0.5349    0.5750         0    0.5743
    0.5716    0.5676    0.4705    0.3750    0.3125    0.3617    0.5743         0



La primera fila si coincide, pero hay otras como la ultima que no coinciden. Deben usar otra version de l similaridad de Gower u otra transformacion de similaridad a distancia.

## Similaridad de Gower-Mahalanobis

In [ ]:
def Matrix_Gower_Maha_Similarity(Data, p1, p2, p3):

    Data_pd = Data

    Data = Data.to_numpy()

    n = len(Data)

    M =  np.empty((n , n))

############################################################
    
    G_vector = np.repeat(0.5, p1)

    for k in range(0, p1):

        G_vector[k] = Data[:,k].max() - Data[:,k].min()

############################################################
   
    Quant_Data = Data[: , 0:p1]

    Quant_Data_pd = Data_pd.iloc[: , 0:p1]

    Binary_Data_pd = Data_pd.iloc[: , (p1):(p1+p2)]

    Multiple_Data = Data[: , (p1+p2):(p1+p2+p3)]

############################################################

    a, b, c, d, p = a_b_c_d_Matrix(Binary_Data_pd)

############################################################

    S_inv = np.linalg.inv( np.cov(Quant_Data , rowvar=False) ) 

    D_maha = Matrix_Dist_Mahalanobis(Data=Quant_Data_pd) 

    Max_D_maha = D_maha.max()

############################################################

    for i in range(0, n):

         for r in range(0, n):

            if i > r :
               
                 M[i,r] = 0
            
            elif i == r :
               
                 M[i,r] = 1
            
            else :

                # Calculo de la distancia de Gower entre las observaciones x_i y x_r

                x = (Quant_Data[i,:] - Quant_Data[r,:])

                x = np.array([x])

                numerator_part_1 = p1 - float( np.sqrt( x @ S_inv @ x.T ) ) / Max_D_maha  

                numerator_part_2 = a[i,r] + alpha(Multiple_Data[i,:], Multiple_Data[r,:])

                numerator = numerator_part_1 + numerator_part_2
 
                denominator = p1 + (p2 - d[i,r]) + p3

                M[i,r] = numerator / denominator
                      
   
    return M  

In [ ]:
M_Gower_Maha = Matrix_Gower_Maha_Similarity(Data=Data_mixed, p1=6, p2=4, p3=3)

M_Gower_Maha

array([[1.        , 0.78829544, 0.70814829, ..., 0.7749854 , 0.7156049 ,
        0.79387387],
       [0.        , 1.        , 0.79624421, ..., 0.97267093, 0.80438259,
        0.68820406],
       [0.        , 0.        , 1.        , ..., 0.80217845, 0.73386904,
        0.62006094],
       ...,
       [0.        , 0.        , 0.        , ..., 1.        , 0.80011792,
        0.67721701],
       [0.        , 0.        , 0.        , ..., 0.        , 1.        ,
        0.62896601],
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        1.        ]])

In [ ]:
M_Gower_Maha = M_Gower_Maha + M_Gower_Maha.T - np.diag(np.repeat(1 , len(M_Gower_Maha)), k=0)

M_Gower_Maha

array([[1.        , 0.78829544, 0.70814829, ..., 0.7749854 , 0.7156049 ,
        0.79387387],
       [0.78829544, 1.        , 0.79624421, ..., 0.97267093, 0.80438259,
        0.68820406],
       [0.70814829, 0.79624421, 1.        , ..., 0.80217845, 0.73386904,
        0.62006094],
       ...,
       [0.7749854 , 0.97267093, 0.80217845, ..., 1.        , 0.80011792,
        0.67721701],
       [0.7156049 , 0.80438259, 0.73386904, ..., 0.80011792, 1.        ,
        0.62896601],
       [0.79387387, 0.68820406, 0.62006094, ..., 0.67721701, 0.62896601,
        1.        ]])

In [ ]:
def Matrix_Dist_Gower_Mahalanobis(Data, p1, p2, p3):

    M = Matrix_Gower_Maha_Similarity(Data, p1, p2, p3)

    M = M + M.T - np.diag(np.repeat(1 , len(M)), k=0)

    M = np.sqrt( 1 - M )

    return M

In [ ]:
M_Gower_Maha_Distance = Matrix_Dist_Gower_Mahalanobis(Data=Data_mixed, p1=6, p2=4, p3=3)

M_Gower_Maha_Distance

array([[0.        , 0.46011364, 0.54023301, ..., 0.47435704, 0.53328707,
        0.45401116],
       [0.46011364, 0.        , 0.45139317, ..., 0.16531507, 0.44228657,
        0.55838691],
       [0.54023301, 0.45139317, 0.        , ..., 0.44477135, 0.51587882,
        0.61639197],
       ...,
       [0.47435704, 0.16531507, 0.44477135, ..., 0.        , 0.44708173,
        0.56813994],
       [0.53328707, 0.44228657, 0.51587882, ..., 0.44708173, 0.        ,
        0.60912559],
       [0.45401116, 0.55838691, 0.61639197, ..., 0.56813994, 0.60912559,
        0.        ]])

# Distancia de Gower generalizada: sin usar matrices de Gram

### Variabilidad geométrica

In [139]:
def var_geom(M_2_Distance):

    n = len(M_2_Distance)

    VG = (1/(2*n**2))*np.sum(M_2_Distance)

    # np.sum(M_2_Distance) es la suma de todos los elementos de la matriz M_2_Distance

    return VG

In [170]:
def Generalized_Gower_Distance(Data, p1, p2, p3, d1, d2, d3, Related_Metric_Scaling=False) :


    n = len(Data)

    M =  np.empty((n , n))

 
    Quant_Data = Data.iloc[: , 0:p1]

    Binary_Data = Data.iloc[: , (p1):(p1+p2)]

    Multiple_Data = Data.iloc[: , (p1+p2):(p1+p2+p3) ]


    # Para que el siguiente paso funciones es necesario que las funciones con las 
    # que vamos a calcular las matrices de distancias tengan un nombre con el siguiente
    # formato: Matrix_Dist_d  donde d = Euclidea, Minkowski, Mahalanobis, Sokal, Jaccard, Simple_Matching etc


    Matrix_Dist_d1 = globals()["Matrix_Dist_" + d1]
    Matrix_Dist_d2 = globals()["Matrix_Dist_" + d2]
    Matrix_Dist_d3 = globals()["Matrix_Dist_" + d3]

   
    D1 = Matrix_Dist_d1(Quant_Data)
    D2 = Matrix_Dist_d2(Binary_Data)
    D3 = Matrix_Dist_d3(Multiple_Data)

    D1_2 = np.square(D1)
    D2_2 = np.square(D2)
    D3_2 = np.square(D3)


    VG1 = var_geom(D1_2)
    VG2 = var_geom(D2_2)
    VG3 = var_geom(D3_2)


    D1_2_standard = D1_2 / VG1
    D2_2_standard = D2_2 / VG2
    D3_2_standard = D3_2 / VG3


    if Related_Metric_Scaling == False :
                
        D_2 = D1_2_standard + D2_2_standard + D3_2_standard 

        D = np.sqrt(D_2)


    elif Related_Metric_Scaling == True :

        ones = np.ones((n, 1))
        ones_T = np.ones((1, n))

        H = np.identity(n) - (1/n)*(ones @ ones_T)

        G_1 = -(1/2)*(H @ D1_2_standard @ H)
        G_2 = -(1/2)*(H @ D2_2_standard @ H)
        G_3 = -(1/2)*(H @ D3_2_standard @ H)

        G_1_sqrt = np.sqrt(G_1)
        G_2_sqrt = np.sqrt(G_2)
        G_3_sqrt = np.sqrt(G_3)

        G = G_1+G_2+G_3 - (1/3)*( G_1_sqrt @ G_2_sqrt + G_1_sqrt @ G_3_sqrt +
                                      G_2_sqrt @ G_1_sqrt + G_2_sqrt @ G_3_sqrt +
                                      G_3_sqrt @ G_1_sqrt + G_3_sqrt @ G_2_sqrt )
            
        g = np.diag(G)                     

        D_2 = g @ ones_T + ones @ g - 2*G

        D = np.sqrt(D_2)

    return D , D_2

In [171]:
D, D_2 = Generalized_Gower_Distance(Data=Data, p1=6, p2=4, p3=3, d1='Euclidea', d2='Sokal', d3='Matching', Related_Metric_Scaling=False)

In [172]:
D

array([[0.        , 1.89001492, 3.89851824, ..., 2.74001795, 2.54939753,
        2.95794239],
       [0.        , 0.        , 3.4273224 , ..., 1.98292844, 1.75531238,
        3.52969462],
       [0.        , 0.        , 0.        , ..., 2.80281931, 3.02532799,
        4.74663048],
       ...,
       [0.        , 0.        , 0.        , ..., 0.        , 1.77002666,
        4.05462793],
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        3.66020899],
       [0.        , 0.        , 0.        , ..., 0.        , 0.        ,
        0.        ]])

In [173]:
D_2

array([[ 0.        ,  3.57215639, 15.1984445 , ...,  7.50769838,
         6.49942777,  8.74942321],
       [ 0.        ,  0.        , 11.74653881, ...,  3.9320052 ,
         3.08112154, 12.45874413],
       [ 0.        ,  0.        ,  0.        , ...,  7.8557961 ,
         9.15260945, 22.53050092],
       ...,
       [ 0.        ,  0.        ,  0.        , ...,  0.        ,
         3.13299439, 16.44000763],
       [ 0.        ,  0.        ,  0.        , ...,  0.        ,
         0.        , 13.39712985],
       [ 0.        ,  0.        ,  0.        , ...,  0.        ,
         0.        ,  0.        ]])

In [174]:
D, D_2 = Generalized_Gower_Distance(Data=Data, p1=6, p2=4, p3=3, d1='Euclidea', d2='Sokal', d3='Matching', Related_Metric_Scaling=True)

In [ ]:
D

In [ ]:
D_2

# Distancia de Gower generalizada: usando matrices de Gram

# Distancia de Mahalanobis Robusta

## Estimacion Robusta de la varianza

### Metodo 1: MAD

### Metodo 2: Trimmed variance

### Metodo 3: Winsorized variance


## Estimacion Robusta de la matriz de covarianzas

### Algoritmo para lograr que la estimacion robusta de la matriz de covarianzas sea definida positiva